docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+   robin, or capacity based).
+1. A tenant which is now attached to node 1 will _also_ still be attached to node
+   0, from the perspective of node 0. Let's say it was originally attached with
+   generation suffix 00000001-0000-00000001 -- its new suffix on node 1
+   might be 00000002-0001-00000001 (attachment gen 2, on node 1).
+1. S3 writes will continue from nodes 0 and 1: there will be an index*part.json-00000001-0000-00000001
+   \_and* an index_part.json-00000002-0001-00000001. Objects written under the old suffix
+   after the new attachment was created do not matter from the rest of the system's
+   perspective: the endpoints are reading from the new attachment location. Objects
+   written by node 0 are just garbage that can be cleaned up at leisure.
+
+Eventually, node 0 will somehow realize it should stop running by some means:
+
+- A hard-stop of the VM it is running on
+- It tries to communicate with another peer that sends it an error response
+  indicating its generation is out of date
+- It tries to do some deletions, and discovers when synchronizing with the
+  control plane that its generation number is stale.
+
+#### Case B: direct node replacement with blank node (cold standby)
+
+1. Let's say node 0 fails, and there may be some other peers but they aren't relevant.
+1. Some external mechanism notices that the node is unavailable, and creates
+   a "new node 0" (Node 0b) which is a physically separate server. The original node 0
+   (Node 0a) may still be running.
+1. On startup, node 0b acquires a new generation number. It doesn't have any local state,
+   so the control plane must send attach requests to node 0b for all the tenants. However,
+   these do not have to increment the attachment generation as they're still attached to
+   logical node 0, it's just a different physical node. So we have an O(tenant_count) communication
+   to the pageserver, but not an O(tenant_count) write to the database.
+1. S3 writes continue from nodes 0a and 0b, but the writes do not collide due to different
+   node generation in the suffix, and the writes from node 0a are not visible to the rest
+   of the system because endpoints are reading only from node 0b.
+
+# Appendix B: interoperability with other features
+
+## Sharded Keyspace
+
+The design in this RFC maps neatly to a sharded keyspace design where subsets of the key space
+for a tenant are assigned to different pageservers:
+
+- the "unit of work" for attachments becomes something like a TenantShard rather than a Tenant
+- TenantShards get generation numbers just as Tenants do.
+- Write workload (ingest, compaction) for a tenant is spread out across pageservers via
+  TenantShards, but each TenantShard still has exactly one valid writer at a time.
+
+## Read replicas
+
+For historical reads to LSNs below the remote persistent LSN, any node may act as a reader at any
+time: remote data is logically immutable data, and the use of deferred deletion in this RFC helps
```

Comment:
IMO deferred deletion is orthogonal to gc process that is in charge of preserving historical data.

Reply:
I agree -- maybe I'm using some overloaded terms here?  When I talk about deletion here, I just mean the physical deletion of objects, rather than the logic about what logical data/lsns may be deleted.